### PR TITLE
start chord header tasks as soon as possible

### LINKED ISSUE #3021: group and chord: subtasks are not executed until entire generator is consumed
When passing a generator to `group` or `chord`, sub-tasks are not submitted to the broker for execution until the entire generator is consumed.  The expected result is that subtasks are submitted during iteration, as soon as they are visited.

Here is the test code I'm using:

``` python
from celery import Celery, group
from celery.utils.log import get_task_logger

app = Celery(__name__, broker='redis://localhost:6379/0',
             backend='redis://localhost:6379/0')

logger = get_task_logger(__name__)

def longyield():
    import time
    for i in xrange(5):
        time.sleep(2)
        print "yielding", i
        yield add.s(i, i)

@app.task
def add(x, y):
    logger.info("adding {} + {}".format(x, y))
    return x + y

def itertest():
    it = longyield()
    g = group(it)
    result = g.delay()
    print result.get()
```

I'm testing with 4.0rc1.

### PR DESCRIPTION
**Pull request by mjhoffman65**
Fixes #3021 . Previously, header tasks were not submitted to the broker for
execution until the entire generator was consumed.

Continuation of https://github.com/celery/celery/pull/6112 and much of this was originally pulled from https://github.com/celery/celery/pull/3043.

Update: this was split into 2 PRs per below review: https://github.com/celery/celery/pull/6589 should be merged first.  This has been rebased on top of that branch.

### CODE REVIEW

File: celery/canvas.py.1
Comment (auvipy) [inline celery/canvas.py.1:4]: curious what this file doing? just renamed?

Comment (mjhoffman65) [reply celery/canvas.py.1]: Oops, this shouldn't have been included.  It's should be gone now.

**Review by auvipy — CHANGES_REQUESTED** [review]
I must have to say great effort so far, could you check this PR https://github.com/celery/celery/pull/6542 and split the related part to another PR and after that get merged rebase this and review & test?

**Comment by mjhoffman65** [comment]
> I must have to say great effort so far, could you check this PR #6542 and split the related part to another PR and after that get merged rebase this and review & test?

Thanks!  Just to make sure I understand, split this PR into 2: first one will replace #6542 and include only the `_regen` changes and the second one will include the rest?

**Comment by auvipy** [comment]
> > I must have to say great effort so far, could you check this PR #6542 and split the related part to another PR and after that get merged rebase this and review & test?
> 
> @auvipy Thanks! Just to make sure I understand, split this PR into 2: first one will replace #6542 and include only the `_regen` changes and the second one will include the rest?

yes.

**Comment by auvipy** [comment]
also, check the conflicts please

**Comment by thedrow** [comment]
I think we had another PR for this one, didn't we? @auvipy 

@mjhoffman65 Thank you very much for the contribution.
Would you kindly rebase this so I can review?

**Comment by mjhoffman65** [comment]
@auvipy @thedrow Merge conflicts have been resolved and this PR has been split (https://github.com/celery/celery/pull/6589) as requested.

File: celery/utils/functional.py
Comment (thedrow) [inline celery/utils/functional.py:235]: Isn't this line going to be a problem?
We'll lose an item while checking if it exists.
Shouldn't we also add it to `self. __consumed`?

**Review by thedrow — CHANGES_REQUESTED** [review]
I need to review this in detail.
In the meantime, I suspect one thing has to be changed.

File: celery/utils/functional.py
Comment (mjhoffman65) [reply celery/utils/functional.py:235]: Added a test to make sure we aren't losing any items: https://github.com/celery/celery/pull/6576/commits/f0d4f2e8915dd5b2710c372c5454218ab89e08db.  That next call drops into `self.__iter__` which will add to `self.__consumed`.

**Comment by auvipy** [comment]
rebase please

**Comment by mjhoffman65** [comment]
> rebase please

:+1: rebased

**Comment by maybe-sybr** [comment]
A quick look at the changes to the redis backend make me think that we can't land this in a patch release since it introduces new key dependencies when handling chords in flight.

I'm yet to review in full.

**Review by thedrow — CHANGES_REQUESTED** [review]

**Comment by thedrow** [comment]
@maybe-sybr Feel free to move the milestone to 5.1.0 if you deem it necessary.

File: celery/canvas.py
Comment (auvipy) [inline celery/canvas.py:168]: do the changes related to chord_size going to break the current API?

File: celery/canvas.py
Comment (auvipy) [inline celery/canvas.py:1175]: so we are going to use chord_length instead of chord_size?

Comment (thedrow) [reply celery/canvas.py:1175]: Does chord_size appear anywhere else? I remember that it does but I'm not sure.
Let's try to keep the same terminology.

File: celery/canvas.py
Comment (auvipy) [inline celery/canvas.py:268]: @thedrow here is another

File: celery/canvas.py
Comment (auvipy) [inline celery/canvas.py:533]: are we spliting chord_size with chord_length & chord_size_bytes?

File: celery/canvas.py
Comment (mjhoffman65) [reply celery/canvas.py:168]: I don't think it should break anything and the tests are passing for me.  Is there a missing test case I should add to make sure the API is ok?

File: celery/canvas.py
Comment (mjhoffman65) [reply celery/canvas.py]: Changed `chord_length` back to `chord_size` to keep the same terminology.

File: celery/canvas.py
Comment (mjhoffman65) [reply celery/canvas.py:533]: Removed the references to `chord_length` to avoid confusion.  Here `chord_size` is removed from the chord object since it is set on the redis backend instead.  

`chord_size_bytes` exists because redis returns a integer back as bytes.
See https://github.com/andymccurdy/redis-py/blob/1a41cfd95a53a95b078084d8627be6b6fba3bb71/tests/test_commands.py#L219

Comment (auvipy) [reply celery/canvas.py:533]: cool thanks for clearing

**Comment by auvipy** [comment]
I was working on fixing the CI and running unit test got this error in CI

t/unit/utils/test_functional.py::test_regen::test_nonzero__does_not_consume_more_than_first_item FAILED [ 66%]

=================================== FAILURES ===================================
________ test_regen.test_nonzero__does_not_consume_more_than_first_item ________

self = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>

>   ???
E   assert not True
E    +  where True = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>.consumed_second_item

t/unit/utils/test_functional.py:145: AssertionError

**Comment by mjhoffman65** [comment]
> I was working on fixing the CI and running unit test got this error in CI
> 
> t/unit/utils/test_functional.py::test_regen::test_nonzero__does_not_consume_more_than_first_item FAILED [ 66%]
> 
> =================================== FAILURES ===================================
> ________ test_regen.test_nonzero__does_not_consume_more_than_first_item ________
> 
> self = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>
> 
> > ???
> > E   assert not True
> > E    +  where True = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>.consumed_second_item
> 
> t/unit/utils/test_functional.py:145: AssertionError

Which branch are you seeing this test failure on?  I'm having trouble recreating it

**Comment by mjhoffman65** [comment]
(Rebased off of master)

**Comment by auvipy** [comment]
https://github.com/celery/celery/pull/6587/checks?check_run_id=1749550836

**Comment by mjhoffman65** [comment]
> https://github.com/celery/celery/pull/6587/checks?check_run_id=1749550836

Thanks!  Looking into it

**Comment by auvipy** [comment]
pull from master again plz

**Comment by mjhoffman65** [comment]
```
_______________ ERROR collecting t/unit/utils/test_functional.py _______________
ImportError while importing test module '/home/runner/work/celery/celery/t/unit/utils/test_functional.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
t/unit/utils/test_functional.py:2: in <module>
    from celery.utils.functional import (DummyContext, first, firstmethod,
E   ImportError: cannot import name 'lookahead'
=========================== short test summary info ============================
ERROR t/unit/utils/test_functional.py
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
========================= 11 skipped, 1 error in 4.63s =========================
```
@auvipy The import error on the tests makes me think these checks might not be using my branch?  The `lookahead` was introduced in this branch.

**Comment by auvipy** [comment]
because that is not yet merged into master branch yet :D

**Comment by auvipy** [comment]
ops sorry that looks weired

**Review by auvipy — CHANGES_REQUESTED** [review]
can you send this unit test fix to a separate PR to check the CI?

**Comment by mjhoffman65** [comment]
@auvipy https://github.com/celery/celery/pull/6613

File: celery/canvas.py
Comment (auvipy) [inline celery/canvas.py:1047]: can we add some comment here explaining the change?

Comment (mjhoffman65) [reply celery/canvas.py:1047]: `_upgrade` used to just set the chord size which has been refactored out since it is set on the backend now instead.
image

Comment (auvipy) [reply celery/canvas.py:1047]: btw do you have the bandwidth to check the pypy failures? maybe in a different PR. also as we don't have CI for Integration tests now, did all the integration tests pass locally?

Comment (mjhoffman65) [reply celery/canvas.py:1047]: yep, I can take a look at the pypy failures in a separate PR.  Integration tests are all passing locally

Comment (mjhoffman65) [reply celery/canvas.py:1047]: Here's a PR I started to look into the PyPy failures: https://github.com/celery/celery/pull/6635.  Need some feedback on it when you get some time @auvipy.

**Comment by mjhoffman65** [comment]
Is there anything I can do to move this forward?  I'm happy to help out with the build separately, but also eager to get these changes into master.

**Comment by auvipy** [comment]
> Is there anything I can do to move this forward? I'm happy to help out with the build separately, but also eager to get these changes into master.

this will surely go into celery 5.1, we just have to see if we are going to release 5.0.6 before that.

**Comment by thedrow** [comment]
We're not going to release 5.0.6. Instead this will land in 5.1.0.

File: celery/canvas.py
Comment (thedrow) [reply celery/canvas.py:168]: If the integration tests will pass, everything is fine and this can be removed. Can you run them to make sure?

File: celery/canvas.py
Comment (thedrow) [inline celery/canvas.py:1169]: We should probably add comments here.

File: celery/canvas.py
Comment (thedrow) [inline celery/canvas.py:1222]: Please add a comment explaining that we're draining from a generator.

File: celery/canvas.py
Comment (thedrow) [inline celery/canvas.py:1251]:
```suggestion
            yield from (task.freeze(group_id=group_id,
                              chord=chord,
                              root_id=root_id,
                              parent_id=parent_id,
                              group_index=group_index)
                              for group_index, task in enumerate(tasks))
```
Isn't this neater?

File: celery/canvas.py
Comment (thedrow) [inline celery/canvas.py:1256]:
```suggestion
       yield from (maybe_signature(task, app=self._app).clone() for task in tasks)
```

Isn't this neater?

File: celery/canvas.py
Comment (thedrow) [inline celery/canvas.py:1474]: Was this a different bug?

File: celery/utils/functional.py
Comment (thedrow) [inline celery/utils/functional.py:172]: I just learned next() has a default. Thanks!

**Review by thedrow — CHANGES_REQUESTED** [review]
Overall, I just have some minor comments.
This is a very good job. Well done and thank you for doing this.

**Comment by thedrow** [comment]
@mjhoffman65 If you'd like to contribute more, we can add you to our team.

**Comment by auvipy** [comment]
> @mjhoffman65 If you'd like to contribute more, we can add you to our team.

I think it's still to early.

**Review by thedrow — CHANGES_REQUESTED** [review]
Can you please take a look into the test failure?

**Comment by mjhoffman65** [comment]
> Can you please take a look into the test failure?

Should be fixed now

File: celery/canvas.py
Comment (mjhoffman65) [reply celery/canvas.py:1173]: Added a comment here.

File: celery/canvas.py
Comment (mjhoffman65) [reply celery/canvas.py:1227]: Added a comment.

File: celery/canvas.py
Comment (mjhoffman65) [reply celery/canvas.py:1477]: Creating a GroupResult will drain the header generator, so I had to refactor it out of `apply_chord`. This line is the last piece of that refactoring.

File: celery/canvas.py
Comment (mjhoffman65) [reply celery/canvas.py:168]: Ran the integration tests with docker and I am getting the same results on this branch as I do on the master branch (1 failure that is unrelated to these changes).

image

**Comment by mjhoffman65** [comment]
> > @mjhoffman65 If you'd like to contribute more, we can add you to our team.
> 
> I think it's still too early.

I'd love to contribute more.  No worries about joining the team if it is too early.  But let me know what I can help on next.

**Review by auvipy — APPROVED** [review]

File: celery/canvas.py
Comment (thedrow) [reply celery/canvas.py:1477]: Got it.

**Review by thedrow — APPROVED** [review]
LGTM. Thank you!

**Comment by auvipy** [comment]
thanks a lot to all of you who were involved to make this PR merge <3

File: celery/canvas.py
Comment (naomielst) [inline celery/canvas.py:1229]: Hi @mjhoffman65 , I'm a new project member. 
I am trying to understand what you did here, I would appreciate if you could clarify it for me. 
I don't get why you zipped `tasks2` with `results` only to get `x[0]`, resulting in a generator with only `tasks2`'s tasks. Why did you write this and not `self.tasks = regen(tasks2)`?
TIA :)

**Pull request merged**
